### PR TITLE
[SEDONA-30] Add raster data support in Sedona SQL

### PR DESCRIPTION
**Pull request by shantanuaggarwal2695**
## Is this PR related to a proposed Issue?
https://issues.apache.org/jira/projects/SEDONA/issues/SEDONA-30

## What changes were proposed in this PR?
Three different constructors have been added :
1. ST_GeomFromRaster  - It fetches polygonal coordinates from raster image
2. ST_GeomWithBandsFromGeoTiff - It constructs a data frame from a raster image with a fixed schema (Geometry and band values)
3. ST_GetBand - Fetches a particular band value from the result of ST_DataframeFromRaster

## How was this patch tested?
Patch has been tested using the Hadoop mini-cluster. Some more test cases can be added

## Did this PR include necessary documentation updates?
Yes

### CODE REVIEW

File: core/pom.xml [hunk 1]
```diff
@@ -60,6 +60,11 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
```
Comment (jiayuasu) [inline core/pom.xml:63]: Please move all geotools dependency to the parent pom. See here: https://github.com/apache/incubator-sedona/blob/master/pom.xml#L120

Make sure you use the geotools scope variable for the scope

File: pom.xml [hunk 2]
```diff
@@ -112,6 +112,13 @@
             <version>0.1.0</version>
             <scope>${dependency.scope}</scope>
         </dependency>
+        <dependency>
+            <groupId>org.geotools</groupId>
+            <artifactId>gt-geotiff</artifactId>
+            <version>${geotools.version}</version>
```
Comment (jiayuasu) [inline pom.xml:118]: Same here

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala [hunk 3]
```diff
@@ -26,7 +26,7 @@ import org.apache.spark.sql.catalyst.expressions.Expression
 import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
 import org.apache.spark.sql.catalyst.util.GenericArrayData
 import org.apache.spark.sql.sedona_sql.UDT.GeometryUDT
-import org.apache.spark.sql.types.{DataType, Decimal}
+import org.apache.spark.sql.types._
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala:29]: If this change is not necessary, do not commit this change. You can replace the "_" with the original content

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala [hunk 4]
```diff
@@ -34,14 +34,13 @@ import org.geotools.geometry.jts.JTS
 import org.geotools.referencing.CRS
 import org.locationtech.jts.algorithm.MinimumBoundingCircle
 import org.locationtech.jts.geom.{PrecisionModel, _}
-import org.locationtech.jts.io.WKBWriter
+import org.locationtech.jts.linearref.LengthIndexedLine
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala:37]: Same here. If this change is not necessary, do not commit this change

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    var coverage:GridCoverage2D = null
+
+    try coverage = reader.read(null)
+    catch {
+      case giveUp: IOException =>
+        throw new RuntimeException(giveUp)
+    }
+    reader.dispose()
+    val source = coverage.getCoordinateReferenceSystem
+    val target = CRS.decode("EPSG:4326", true)
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala:73]: why do you implement CRS transformation here? This should be handled by Sedona ST_Transform outside this funciton.

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    mySchema
+  }
+
+  override def dataType: DataType = getSchema()
+
+  override def children: Seq[Expression] = inputExpressions
+}
+
+//  get a particular band from a raster dataframe
+case class ST_getBand(inputExpressions: Seq[Expression])
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala:217]: Use ST_GetBand

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+    try coverage = reader.read(null)
+    catch {
+      case giveUp: IOException =>
+        throw new RuntimeException(giveUp)
+    }
+    reader.dispose()
+    val source = coverage.getCoordinateReferenceSystem
+    val target = CRS.decode("EPSG:4326", true)
+    val targetCRS = CRS.findMathTransform(source, target)
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala:131]: If CRS transformation is not necessary here,  don't do it then.

Comment (jiayuasu) [reply sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala:131]: It looks like you are reading geom and bands together in ST_DataframeFromRaster. A better name could be ST_GeomWithBandsFromGeoTiff

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    // This is an expression which takes one input expressions
+    assert(inputExpressions.length == 2)
+    val geomString = inputExpressions(0).eval(inputRow).asInstanceOf[UTF8String].toString
+    val totalBands = inputExpressions(1).eval(inputRow).asInstanceOf[Int]
+    val geometry = readGeometry(geomString)
+    val bandvalues = getBands(geomString, totalBands).toArray
+    returnValue(geometry.toGenericArrayData,bandvalues, 2)
+  }
+
+  private def readGeometry(url: String): Geometry = {
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala:116]: Please reduce the duplicate code by merging them into a function and re-use it in different places

**Review by jiayuasu — CHANGES_REQUESTED** [review]
See my comments above

**Comment by shantanuaggarwal2695** [comment]
@jiayuasu  I have tried to fix all the comments

File: core/pom.xml (hunk 1 above)
Comment (shantanuaggarwal2695) [reply core/pom.xml]: Fixed

File: pom.xml (hunk 2 above)
Comment (shantanuaggarwal2695) [reply pom.xml:118]: Fixed

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala (hunk 3 above)
Comment (shantanuaggarwal2695) [reply sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala]: Fixed

File: sql/src/main/scala/org/apache/sedona/sql/UDF/Catalog.scala [hunk 9]
```diff
@@ -81,7 +81,10 @@ object Catalog {
     ST_IsRing,
     ST_FlipCoordinates,
     ST_LineSubstring,
-    ST_LineInterpolatePoint
+    ST_LineInterpolatePoint,
+    ST_GeomFromRaster,
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/sedona/sql/UDF/Catalog.scala:85]: Please change the name to ST_GeomFromGeoTiff

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala (hunk 4 above)
Comment (jiayuasu) [reply sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala:37]: Please don't commit any change to this file

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+    val h: Int = maxDimensions.getCoordinateValue(1) + 1
+    val numBands: Int = bands
+
+    val bandValues: util.List[util.List[Double]] = new util.ArrayList[util.List[Double]](numBands)
+
+    for (i <- 0 until numBands) {
+      bandValues.add(new util.ArrayList[Double])
+    }
+
+    for (i <- 0 until w) {
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala:147]: I believe this loop and the next loop can be merged

File: sql/src/test/java/org/apache/sedona/sql/rasterTestJava.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
+package org.apache.sedona.sql;
```
Comment (jiayuasu) [inline sql/src/test/java/org/apache/sedona/sql/rasterTestJava.java:1]: Please move this test to Scala test. Don't write it in Java

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala [hunk 12]
```diff
@@ -303,5 +303,4 @@ trait UserDataGeneratator {
     }
     return userData
   }
-}
-
+}
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Constructors.scala:306]: Please don't commit any change to this file

File: core/pom.xml [hunk 13]
```diff
@@ -60,6 +60,7 @@
                 </exclusion>
             </exclusions>
         </dependency>
+
```
Comment (jiayuasu) [inline core/pom.xml:63]: please don't commit any change to this file

**Review by jiayuasu — CHANGES_REQUESTED** [review]
In addition to my comments above, please also add docs for your Constructors.
https://github.com/apache/incubator-sedona/edit/master/docs/api/sql/GeoSparkSQL-Constructor.md

Please follow this: http://sedona.apache.org/download/compile/#compile-the-doc-website

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala (hunk 10 above)
Comment (shantanuaggarwal2695) [reply sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala:147]: @jiayuasu  The current loop returns an array of size equal to the number of bands for every pixel and I use it to construct a 2D array of size number of bands * number of pixels while the second loop flattens this array. If I need to merge both the loops, I have to keep a separate starting index for every band which may vary.

**Review by jiayuasu — APPROVED** [review]

**Pull request merged**